asynclocalstorage.md

```diff
@@ -0,0 +1,371 @@
+# AsyncLocalStorage - Web-interoperable Runtimes portable subset
+
+This document describes a portable subset of the Node.js `AsyncLocalStorage` API
+that can be implemented by non-Node.js runtimes without dependency on the
+lower-level `async_hooks` API.
+
+## Logical model
+
+This subset uses a fundamentally different underlying logical model than what
+is *currently* implemented in Node.js' `AsyncLocalStorage` implementation based
+on `async_hooks`. I won't go into the details of Node.js' current model here,
+however. Instead, I'll focus on the logical model assumed by this subset.
+
+At any given time, there is a current `Async Context Frame` (which, for ease
+of referencing I'll just call `frame` from here on out).
+
+The `frame` is effectively an *immutable* map of `storage keys` to `store`
+value. We refer to this as the `frame`'s `storage context`. A `storage key`
+is associated with exactly one `AsyncLocalStorage` object instance.
+
+There is always a `root frame` whose `storage context` is always empty.
+
+Whenever a new value is associated with a given `storage key`, a new
+`frame` is created with the `value` associated with that `storage key`
+set in the new frame's storage context.
+
+Whenever a new `frame` is created, it inherits a *copy* of the current
+`frame`'s `storage context` before setting the new `value` associated
+with the storage key.
+
+In **pseudo-code**, this looks something like:  (this is **NOT** the API,
+this is a pesudo-code illustration of the abstract model)
+
+```js
+class AsyncLocalStorage {
+  #storageKey = {}; // the storage key is opaque and unique to this
+                    // ALS instance.
+  run(store, fn) {
+    return AsyncContextFrame.run(#storageKey, store, fn);
+  }
+  
+  getStore() {
+    return AsyncContextFrame::current().get(this.#storageKey);
```

Comment:
Nit: What's with the `::`? The rest of this is JS; why not use `.` here?

Reply:
Heh, C++ brain invading JavaScript brain.